Support write to buffer api for SerializedFileWriter

zhuqi-lucas opened the pull request:
# Which issue does this PR close?

Currently, no pub api to support write the internal buffer for SerializedFileWriter, it's very helpful when we want to add low level API for example:
- https://github.com/apache/datafusion/issues/16374
- https://github.com/apache/datafusion/pull/16395

Because that we want to update the buf bytes written, if we use the buf internal file to write, we can't update the internal buf written bytes.

The consistent update for the bytes written metrics is the key for our custom index write.


# Rationale for this change

Add API to support write with buf byteswritten updating.

# What changes are included in this PR?

Add API to support write with buf byteswritten updating.

# Are there any user-facing changes?
No

If there are user-facing changes then we may require documentation to be updated before approving the PR.

If there are any breaking changes to public APIs, please call them out.

etseidl reviewed (approved):
Thanks, seems sensible. I think the documentation for `SerializedFileWriter` could stand some improving to highlight _why_ it's a bad idea to write directly to the underlying writer.

zhuqi-lucas commented:
Thank you @etseidl for review, i also add the reminder docs in latest PR for users who want to write directly.

alamb commented on parquet/src/arrow/arrow_writer/mod.rs:
```diff
@@ -320,14 +320,21 @@ impl<W: Write + Send> ArrowWriter<W> {
 
     /// Returns a reference to the underlying writer.
+    ///
+    /// **Warning**: if you write directly to this writer, you will skip
```
Given this is a read only reference (immutable) the warning here is probably not necessary.

zhuqi-lucas replied:
Got it now, thanks @alamb , addressed in latest PR!

alamb commented on parquet/src/arrow/arrow_writer/mod.rs:
```diff
@@ -320,14 +320,21 @@ impl<W: Write + Send> ArrowWriter<W> {
     ///
-    /// It is inadvisable to directly write to the underlying writer, doing so
-    /// will likely result in a corrupt parquet file
+    /// **Warning**: if you write directly to this writer, you will skip
```
Thank you

alamb commented on parquet/src/file/writer.rs:
```diff
@@ -390,13 +390,33 @@ impl<W: Write + Send> SerializedFileWriter<W> {
 
     /// Returns a reference to the underlying writer.
+    ///
+    /// **Warning**: if you write directly to this writer, you will skip
```
Likewise here since it is not possible to write to an immutable buffer we can probably skip this warning

alamb commented on parquet/src/file/writer.rs:
```diff
@@ -390,13 +390,33 @@ impl<W: Write + Send> SerializedFileWriter<W> {
+    /// the file footer’s recorded offsets and sizes to diverge from reality,
+    /// resulting in an unreadable or corrupted Parquet file.
+    ///
+    /// If you want to write safely to the underlying writer, use [`Self::write_all`].
```
❤️

alamb commented on parquet/src/arrow/arrow_writer/mod.rs:
```diff
@@ -320,14 +320,21 @@ impl<W: Write + Send> ArrowWriter<W> {
+    /// resulting in an unreadable or corrupted Parquet file.
     pub fn inner_mut(&mut self) -> &mut W {
         self.writer.inner_mut()
     }
```
I wonder should we add `write_all` to the ArrowWriter too?

alamb commented on parquet/src/file/writer.rs:
```diff
@@ -390,13 +390,33 @@ impl<W: Write + Send> SerializedFileWriter<W> {
         self.buf.inner()
     }
 
+    /// Writes the given buf bytes to the internal buffer.
```
It might be nice to add a note here about why you might do this

For example, something like

```suggestion
    /// Writes the given buf bytes to the internal buffer.
    ///
    /// This can be used to write raw data to an in-progress parquet file, for 
    /// example custom index structures or other payloads. Other parquet readers 
    /// will skip this data when reading the files.
```

zhuqi-lucas replied:
Thank you @alamb good suggestion.

alamb reviewed (approved):
Thank you @zhuqi-lucas and @etseidl 

I have some more comments about comments 😆  but the code looks great here.

Thank you!

zhuqi-lucas commented:
> Thank you @zhuqi-lucas and @etseidl
> 
> I have some more comments about comments 😆 but the code looks great here.
> 
> Thank you!

Thank you @alamb for review, addressed comments and suggestions in latest PR.

zhuqi-lucas commented:
> * I would really like to make sure this one is in [Release arrow-rs / parquet Minor version `55.2.0` (June 2025) #7394](https://github.com/apache/arrow-rs/issues/7394)

Sounds great to include in this release!

alamb reviewed (approved):
let's do it -- thanks again @zhuqi-lucas
